gps/prune.go

```diff
@@ -56,204 +84,172 @@ var (
-			return errors.Wrap(err, "unexpected error while calculating unused packages")
-		}
-
-		pkg = filepath.ToSlash(pkg)
 		if _, ok := imported[pkg]; !ok {
-			unused[pkg] = struct{}{}
+			unused[pkg] = nil
 		}
+	}
 
-		return nil
-	})
-
-	return unused, err
+	return unused
 }
 
-// collectUnusedPackagesFiles returns a slice of all files in the unused packages in projectDir.
-func collectUnusedPackagesFiles(projectDir string, unusedPackages map[string]struct{}) ([]string, error) {
+// collectUnusedPackagesFiles returns a slice of all files in the unused
+// packages based on fsState.
+func collectUnusedPackagesFiles(fsState filesystemState, unusedPackages map[string]interface{}) []string {
+	// TODO(ibrasho): is this useful?
 	files := make([]string, 0, len(unusedPackages))
 
-	err := filepath.Walk(projectDir, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
+	for _, path := range fsState.files {
+		// Keep perserved files.
+		if isPreservedFile(filepath.Base(path)) {
+			continue
 		}
 
-		// Ignore directories.
-		if info.IsDir() {
-			return nil
-		}
-
-		// Ignore preserved files.
-		if isPreservedFile(info.Name()) {
-			return nil
-		}
-
-		pkg, err := filepath.Rel(projectDir, filepath.Dir(path))
-		if err != nil {
-			return errors.Wrap(err, "unexpected error while calculating unused packages")
-		}
+		pkg := filepath.ToSlash(filepath.Dir(path))
 
-		pkg = filepath.ToSlash(pkg)
 		if _, ok := unusedPackages[pkg]; ok {
-			files = append(files, path)
+			files = append(files, filepath.Join(fsState.root, path))
 		}
-
-		return nil
-	})
-
-	return files, err
-}
-
-// pruneNonGoFiles delete all non-Go files existing within baseDir.
-// Files with names that are prefixed by any entry in preservedNonGoFiles
-// are not deleted.
-func pruneNonGoFiles(baseDir string, logger *log.Logger) error {
-	files, err := collectNonGoFiles(baseDir, logger)
-	if err != nil {
-		return errors.Wrap(err, "could not collect non-Go files")
 	}
 
-	if err := deleteFiles(files); err != nil {
-		return errors.Wrap(err, "could not prune Go test files")
-	}
-
-	return nil
+	return files
 }
 
-// collectNonGoFiles returns a slice containing all non-Go files in baseDir.
-// Files meeting the checks in isPreservedFile are not returned.
-func collectNonGoFiles(baseDir string, logger *log.Logger) ([]string, error) {
-	files := make([]string, 0)
-
-	err := filepath.Walk(baseDir, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
+// pruneNonGoFiles delete all non-Go files existing in fsState.
+//
+// Files matching licenseFilePrefixes and legalFileSubstrings are not pruned.
+func pruneNonGoFiles(fsState filesystemState) error {
+	// TODO(ibrasho) detemine a sane capacity
+	toDelete := make([]string, 0, len(fsState.files)/4)
+
+	for _, path := range fsState.files {
+		ext := fileExt(path)
+
+		// Refer to: https://sourcegraph.com/github.com/golang/go/-/blob/src/go/build/build.go#L750
```

Comment:
prefer a direct github link rather than sourcegraph, and link to a hash rather than a branch so that it points to a stable block of code.